src/szk.rs

```diff
@@ -0,0 +1,459 @@
+    pub fn new(
+        typ: T,
+        dst: [u8; 8]
+    ) -> Self {
+        Self {
+            typ,
+            dst,
+            phantom: PhantomData,
+        }
+    }
+
+
+    /// The output length of the underlying FLP.
+    pub fn output_len(&self) -> usize {
+        self.typ.output_len()
+    }
+
+    /// The verifier length of the underlying FLP.
+    pub fn verifier_len(&self) -> usize {
+        self.typ.verifier_len()
+    }
+
+    fn adjust_domsep_tag(&self, usage: u16) -> [u8; 8] {
+        let mut adjusted_dst = self.dst.clone();
+        adjusted_dst[6..8].copy_from_slice(&usage.to_be_bytes());
+        adjusted_dst
+    }
+
+    /// Derive a vector of random field elements for consumption by the FLP prover
+    /// Question: is it ok to have the binder value be zero here? This is different from
+    /// both Prio3 (where it's the number of proofs) and the Mastic paper (where it's empty).
+    fn derive_prove_rands(&self, prove_rand_seed: &Seed<SEED_SIZE>) -> Vec<T::Field> {
+        P::seed_stream(
+            prove_rand_seed,
+            &self.adjust_domsep_tag(DST_PROVE_RANDOMNESS),
+            &[0],
+        )
+        .into_field_vec(self.typ.prove_rand_len())
+    }
+
+    fn derive_joint_rand_part(
+        &self,
+        aggregator_blind: &Seed<SEED_SIZE>,
+        measurement_share: &[T::Field],
+        nonce: &[u8; 16],
+    ) -> Result<Seed<SEED_SIZE>, SzkError> {
+        let mut xof = P::init(
+            aggregator_blind.as_ref(),
+            &self.adjust_domsep_tag(DST_JOINT_RAND_SEED),
+        );
+        // Encode measurement_share (currently an array of field elements) into bytes
+        // and include it in the XOF input.
+        let mut bytes =  vec![0u8; T::Field::ENCODED_SIZE];
+        for field_elem in measurement_share{
+            field_elem.encode(&mut bytes)?;
+            xof.update(&bytes)
+        }
+        xof.update(nonce);
+        Ok(xof.into_seed())
+    }
+
+    fn derive_joint_rand_seed(
+        &self,
+        leader_joint_rand_part: &Seed<SEED_SIZE>,
+        helper_joint_rand_part: &Seed<SEED_SIZE>,
+    ) -> Seed<SEED_SIZE> {
+        let mut xof = P::init(
+            &[0; SEED_SIZE],
+            &self.adjust_domsep_tag(DST_JOINT_RAND_SEED),
+        );
+        xof.update(&leader_joint_rand_part.0);
+        xof.update(&helper_joint_rand_part.0);
+        xof.into_seed()
+    }
+
+    fn derive_joint_rands(
+        &self,
+        leader_joint_rand_part: &Option<Seed<SEED_SIZE>>,
+        helper_joint_rand_part: &Option<Seed<SEED_SIZE>>,
+    ) -> (Seed<SEED_SIZE>, Vec<T::Field>) {
+        if self.typ.joint_rand_len() <= 0 {
+            return (Seed::from_bytes([0; SEED_SIZE]), Vec::<T::Field>::with_capacity(0));
+        };
+        let joint_rand_seed = self.derive_joint_rand_seed(leader_joint_rand_part.as_ref().expect("jr_len > 0"),
+                                                          helper_joint_rand_part.as_ref().expect("jr_len > 0"));
```

Comment:
I think this could be simplified a bit if it only handled the case where joint randomness is used, and its callers passed seeds instead of `Option`s of seeds. It would be nice to remove the branch that returns a zero seed. I think this will ultimately require changing the signature of `query()` to return an Option<Seed> for the joint randomness seed. There again, it'd be nice to not have a case where a seed is all zeros.